frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -390,6 +400,45 @@ function ReaderHighlight:addToMainMenu(menu_items)
             separator = i == #highlight_style,
         })
     end
+    table.insert(menu_items.highlight_options.sub_item_table, {
+        text_func = function()
+            local saved_color = self.view.highlight.saved_color or "yellow"
+            for __, v in ipairs(self.highlight_colors) do
+                if v[2] == saved_color then
+                    return T(_("Highlight color: %1"), string.lower(v[1]))
+                end
+            end
+            return T(_("Highlight color: %1"), saved_color)
+        end,
+        enabled_func = function()
+            return Screen:isColorScreen() and self.view.highlight.saved_drawer ~= "invert"
+        end,
+        callback = function(touchmenu_instance)
+            local saved_color = self.view.highlight.saved_color or "yellow"
+            local radio_buttons = {}
+            for _, v in ipairs(self.highlight_colors) do
+                table.insert(radio_buttons, {
+                    {
+                        text = v[1],
+                        checked = v[2] == saved_color,
+                        provider = v[2],
+                    },
+                })
+            end
+            UIManager:show(require("ui/widget/radiobuttonwidget"):new{
+                title_text = _("Highlight color"),
+                width_factor = 0.5,
+                keep_shown_on_apply = false,
+                radio_buttons = radio_buttons,
+                callback = function(radio)
+                    G_reader_settings:saveSetting("highlight_color", radio.provider)
+                    self.view.highlight.saved_color = radio.provider
+                    UIManager:setDirty(self.dialog, "ui")
```

Comment:
This would have effect only for old highlights that do not have item.color.

Generally, we now have inconsistency in style and color settings.
Style can be set: global default, per document default, per highlight.
Color per document default doesn't exist yet.

Reply:
All highlights made after updating the apk should always have item.color set. It being unset should not be expected for new highlights. We should maybe figure out a way to one-time patch all unset highlights just for consistency, but I don't think it's a huge issue in practice. The setting is mostly to change how future highlights will be seen.

Regarding per document default. Where would that be stored (code example would be helpful)? Also, it seems like most document settings are in the bottom menu, not the top one, so I worry about consistency. However, if style is also per-document and also set in the top menu, well then I guess it would be the same thing.

Reply:
It looks like it is done like this?

set:
```
self.ui.doc_settings:saveSetting("highlight_drawer", self.view.highlight.saved_drawer)
```

get:
```
self.view.highlight.saved_drawer = config:readSetting("highlight_drawer")
```

Reply:
Here is my diff which I think makes sense.
Sadly hold_callback doesn't work in radio buttons, so I would need to manually create a window popup with the submenu. Sigh.

```
diff --git a/frontend/apps/reader/modules/readerhighlight.lua b/frontend/apps/reader/modules/readerhighlight.lua
index b1ddba5ca..3dd472bb7 100644
--- a/frontend/apps/reader/modules/readerhighlight.lua
+++ b/frontend/apps/reader/modules/readerhighlight.lua
@@ -430,12 +430,26 @@ function ReaderHighlight:addToMainMenu(menu_items)
                 width_factor = 0.5,
                 keep_shown_on_apply = false,
                 radio_buttons = radio_buttons,
+                text_func = function()
+                    local text = v[1]
+                    if v[2] == G_reader_settings:readSetting("highlight_color") then
+                        text = text .. "   ★"
+                    end
+                    return text
+                end,
                 callback = function(radio)
-                    G_reader_settings:saveSetting("highlight_color", radio.provider)
-                    self.view.highlight.saved_color = radio.provider
+                    self.view.highlight.saved_drawer = radio.provider
+                    self.ui.doc_settings:saveSetting("highlight_color", self.view.highlight.saved_color)
                     UIManager:setDirty(self.dialog, "ui")
                     if touchmenu_instance then touchmenu_instance:updateItems() end
                 end,
+                hold_callback = function(touchmenu_instance)
+                    G_reader_settings:saveSetting("highlight_color", radio.provider)
+                    UIManager:show(Notification:new{
+                        text = T(_("Default highlight color changed to '%1'."), radio.provider),
+                    })
+                    if touchmenu_instance then touchmenu_instance:updateItems() end
+                end,
             })
         end,
     })
@@ -2160,7 +2174,7 @@ function ReaderHighlight:onReadSettings(config)
     self.view.highlight.saved_drawer = config:readSetting("highlight_drawer")
         or G_reader_settings:readSetting("highlight_drawing_style") or self.view.highlight.saved_drawer
     self.view.highlight.saved_color = config:readSetting("highlight_color")
-        or self.view.highlight.saved_color
+        or G_reader_settings:readSetting("highlight_color") self.view.highlight.saved_color
     self.view.highlight.disabled = G_reader_settings:has("default_highlight_action")
         and G_reader_settings:readSetting("default_highlight_action") == "nothing"
 ```

Reply:
I propose the way we use for the style setting.
Global and document defaults have no immediate visual effect, they are set in the upper menu/submenu.
The highlight personal color to be set in the RadioButton widget called from the popup buttontable,

Reply:
> how to remove any immediate visual effect

Do you need? Once all highlights have the `color` property, there will be no immediate effect.

Reply:
> I guess that needs to be removed as a fallback in the drawing code and instead use a hardcoded value?

I see.
How would you (having a color device) like to see old highlights without the `color` property?
With the default color or with old b/w setting?